### DISCUSSION TITLE
Generating batch code from existing game assets

### DISCUSSION THREAD

**Discussion by eartsar** [evennia/evennia#2297] [Evennia Questions] [answered]
Hi folks. Pardon if this is a repeat question, some low-effort searching didn't manifest in any obvious duplicates.

Context: I have recently joined an established project comprised of a single developer, but a few very capable and trusted builders, with the aim of paying down some of the technical debt accrued by the project. One of the bigger ticket items I've identified is that all of the game assets were created through in-engine commands, and not through batch code under source control.

Having all game assets created via batch code, so that a fresh world can be spun up for any reason (other deploy environments, automated testing, you name it), and as the community has pointed out, better auditing of new assets coming into the game, seems to be the way to go. This leaves us with the task of converting these existing assets into batch code.

While we could do this by hand, I'm hoping that I can find a way to automate it, as doing this by hand would be very painful and time consuming.

Has anyone tackled this problem, or have any idea in how to tackle this issue?

**Comment by Griatch** [comment, marked as answer]
There is no in-built way to automate generation of batchcode, there are just too many possible combinations to make it generic - this is because this most often something you want when converting a database from some other, non-Evennia system. 
You have a little unusual use-case though in that you have already created Evennia code that you want to post-convert. Since you presumably have the matching Typeclasses set up already, it may be easier to make a custom extractor for your particular game, where you loop over the objects and have each generate a `create_object` line in a batchcode file.

But overall, as laid out in the docs, we don't recommend starting to build anything susbstantial in the database until you really have a working game already. It helps as lot ot be able to wipe your db quickly and easily when developing and experimenting with your design. For SQlite3 you can just copy&paste a fresh database file over `mygame/server/evennia.db3` to 'reset' the database back to whatever state you copied the file at.

**Reply by eartsar** [reply]
Appreciate the reply, Griatch. Chances are I will have to go down a custom extraction route (expecting this to be very painful) in the coming months, since there are just too many assets to do by hand. Should I tackle this programmatically, I'll post implementation notes, and the resulting script in this thread in case anyone has the same pain point in the future.

**Comment by strikaco** [comment]
Depending on the nature of your attributes, you might find this function useful for serializing your assets:

https://github.com/evennia/evennia/blob/574d7fbd4f65cae60c534fe592803057d452ed7b/evennia/prototypes/spawner.py#L275-L285

If all your attributes store are primitives, it works pretty well. Mixed results otherwise. 

The output should natively work with the `spawn` command or `prototypes.spawner.spawn` function, but if you're dead-set on generating batch code, it should not be difficult to break down each object further into a series of `spawn <x>`/`set <x>/<attr> = <value>`/`tag <x> = <blah>`/`tel <x> = <y>` commands.

**Reply by Griatch** [reply]
@strikaco It's not a bad idea, although that function is not tested to be used for batch-use this way. :smile: 
That said, don't confuse batch-cmd with batch-code. For this I'd not use batch-cmd (which is the example you give) but instead generate a batch-code file where you can use python calls directly.
